src/bytes_mut.rs

```diff
@@ -644,21 +644,37 @@ impl BytesMut {
 
                     self.ptr = vptr(ptr);
                     self.cap = v.capacity();
+                } else {
```

Comment:
This change ignores the fact that we might be overallocating by a significant amount by not checking if there's space to reclaim first before seeing if we still need to reallocate the underlying `Vec<u8>`.

It seems like we should probably do a general check inside the "is unique" if block where we see if our slice is mapped to the underlying `Vec<u8>`, and if not, do that first: copy our bytes, adjust `self.ptr`/`self.cap` correctly, etc.

After that, _then_ we would have the logic to figure out if we have enough capacity to fulfill the call, and if not, proceed with resizing the original allocation.